Basic support for CSS based Ruby in IFC

anttijk opened the pull request:
#### 44bf0d71007ce85571ea893edef23a9b3935f7a8
<pre>
Basic support for CSS based Ruby in IFC
<a href="https://bugs.webkit.org/show_bug.cgi?id=261468">https://bugs.webkit.org/show_bug.cgi?id=261468</a>
rdar://115370429

Reviewed by Alan Baradlay.

The current Ruby implementation uses special elements and renderers.

This patch implements basic support for CSS based ruby in IFC:
- Adds display property values &apos;ruby&apos;, &apos;block-ruby&apos;, &apos;ruby-base&apos; and &apos;ruby-text&apos; but does not expose them outside the user
agent stylesheet.
- Adds render tree building support for these values.
- Sketches an IFC implementation that is able to generate annotation boxes in the simple case. Layout is not affected.

The feature is behind a flag and is disabled by default.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:

Add a setting.

* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::toCSSValueID):
(WebCore::fromCSSValueID):

Use switch for robustness and readability.

* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeDisplay):

Pass the parsing mode so we can limit the new properties to the UA sheet.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/css/ruby.css: Added.

Add a UA stylesheet that sets the display property values for HTML ruby elements.

* Source/WebCore/html/RubyElement.cpp:
(WebCore::RubyElement::createElementRenderer):

Skip the special case renderer generation if enabled.

* Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
(WebCore::Layout::InlineItemsBuilder::collectInlineItems):

Skip over ruby annotation boxes when collecting inline items.
Ruby bases simply map to inline boxes.

* Source/WebCore/layout/formattingContexts/inline/display/InlineDisplayContentBuilder.cpp:
(WebCore::Layout::InlineDisplayContentBuilder::appendInlineBoxDisplayBox):
(WebCore::Layout::InlineDisplayContentBuilder::appendAssociatedRubyAnnotationBoxIfNeeded):

If an inline box has an associated ruby box it is actually a ruby base.
Generate an annotation box and place it above the base.

* Source/WebCore/layout/formattingContexts/inline/display/InlineDisplayContentBuilder.h:
* Source/WebCore/layout/integration/inline/LayoutIntegrationLineLayout.cpp:
(WebCore::LayoutIntegration::LineLayout::updateRenderTreePositions):

Place the annotation box renderer after layout.

* Source/WebCore/layout/layouttree/LayoutBox.cpp:
(WebCore::Layout::Box::isInlineLevelBox const):
(WebCore::Layout::Box::isInlineBox const):
(WebCore::Layout::Box::isBlockContainer const):
(WebCore::Layout::Box::isRubyAnnotationBox const):
(WebCore::Layout::Box::isInternalRubyBox const):
(WebCore::Layout::Box::associatedRubyAnnotationBox const):

Find the annotation if any.

* Source/WebCore/layout/layouttree/LayoutBox.h:
(WebCore::Layout::Box::isImage const):
(WebCore::Layout::Box::isInternalRubyBox const): Deleted.
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::createFor):

Use generic renderers for ruby display values.

* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::isDisplayBlockType):
(WebCore::RenderStyle::isDisplayInlineType):
* Source/WebCore/rendering/updating/RenderTreeBuilder.cpp:
(WebCore::RenderTreeBuilder::attachInternal):
* Source/WebCore/rendering/updating/RenderTreeBuilderInline.cpp:
(WebCore::RenderTreeBuilder::Inline::splitInlines):

With ruby we may find anonymous inlines when splitting inlines for continuations. Just ignore them, they provide no styling.

* Source/WebCore/rendering/updating/RenderTreeBuilderRuby.cpp:
(WebCore::RenderTreeBuilder::Ruby::findOrCreateParentForStyleBasedRubyChild):

Generate anonymous boxes for ruby so we get structure like

&lt;ruby&gt;&lt;ruby-base&gt;base content&lt;/ruby-base&gt;&lt;ruby-annotation&gt;annotation content&lt;/ruby-annotation&gt;&lt;/ruby&gt;

* Source/WebCore/rendering/updating/RenderTreeBuilderRuby.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::equivalentBlockDisplay):
(WebCore::Style::equivalentInlineDisplay):

Add inlinification function.

(WebCore::Style::shouldInlinifyForRuby):
(WebCore::Style::Adjuster::adjust const):

Inlinify children of ruby boxes.

* Source/WebCore/style/UserAgentStyle.cpp:
(WebCore::Style::UserAgentStyle::ensureDefaultStyleSheetsForElement):

Include ruby.css if enabled.

* Source/WebCore/style/UserAgentStyle.h:

Canonical link: <a href="https://commits.webkit.org/267937@main">https://commits.webkit.org/267937@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/15356e4572522bbfb1fae2ce01e77b0e30da2818

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18129 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18462 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19030 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/19968 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16968 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18326 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21759 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18620 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18932 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18349 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18587 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15783 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20846 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15811 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16536 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/23048 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/15704 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16830 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16707 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20929 "Passed tests") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/17407 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17275 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/14632 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/21484 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16365 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/5267 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4311 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20727 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/22718 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17123 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5114 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

alanbaradlay reviewed (approved):
nice (picture some annotation here)

anttijk committed:
Basic support for CSS based Ruby in IFC
https://bugs.webkit.org/show_bug.cgi?id=261468
rdar://115370429

Reviewed by Alan Baradlay.

The current Ruby implementation uses special elements and renderers.

This patch implements basic support for CSS based ruby in IFC:
- Adds display property values 'ruby', 'block-ruby', 'ruby-base' and 'ruby-text' but does not expose them outside the user
agent stylesheet.
- Adds render tree building support for these values.
- Sketches an IFC implementation that is able to generate annotation boxes in the simple case. Layout is not affected.

The feature is behind a flag and is disabled by default.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:

Add a setting.

* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::toCSSValueID):
(WebCore::fromCSSValueID):

Use switch for robustness and readability.

* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeDisplay):

Pass the parsing mode so we can limit the new properties to the UA sheet.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/css/ruby.css: Added.

Add a UA stylesheet that sets the display property values for HTML ruby elements.

* Source/WebCore/html/RubyElement.cpp:
(WebCore::RubyElement::createElementRenderer):

Skip the special case renderer generation if enabled.

* Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
(WebCore::Layout::InlineItemsBuilder::collectInlineItems):

Skip over ruby annotation boxes when collecting inline items.
Ruby bases simply map to inline boxes.

* Source/WebCore/layout/formattingContexts/inline/display/InlineDisplayContentBuilder.cpp:
(WebCore::Layout::InlineDisplayContentBuilder::appendInlineBoxDisplayBox):
(WebCore::Layout::InlineDisplayContentBuilder::appendAssociatedRubyAnnotationBoxIfNeeded):

If an inline box has an associated ruby box it is actually a ruby base.
Generate an annotation box and place it above the base.

* Source/WebCore/layout/formattingContexts/inline/display/InlineDisplayContentBuilder.h:
* Source/WebCore/layout/integration/inline/LayoutIntegrationLineLayout.cpp:
(WebCore::LayoutIntegration::LineLayout::updateRenderTreePositions):

Place the annotation box renderer after layout.

* Source/WebCore/layout/layouttree/LayoutBox.cpp:
(WebCore::Layout::Box::isInlineLevelBox const):
(WebCore::Layout::Box::isInlineBox const):
(WebCore::Layout::Box::isBlockContainer const):
(WebCore::Layout::Box::isRubyAnnotationBox const):
(WebCore::Layout::Box::isInternalRubyBox const):
(WebCore::Layout::Box::associatedRubyAnnotationBox const):

Find the annotation if any.

* Source/WebCore/layout/layouttree/LayoutBox.h:
(WebCore::Layout::Box::isImage const):
(WebCore::Layout::Box::isInternalRubyBox const): Deleted.
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::createFor):

Use generic renderers for ruby display values.

* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator<<):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::isDisplayBlockType):
(WebCore::RenderStyle::isDisplayInlineType):
* Source/WebCore/rendering/updating/RenderTreeBuilder.cpp:
(WebCore::RenderTreeBuilder::attachInternal):
* Source/WebCore/rendering/updating/RenderTreeBuilderInline.cpp:
(WebCore::RenderTreeBuilder::Inline::splitInlines):

With ruby we may find anonymous inlines when splitting inlines for continuations. Just ignore them, they provide no styling.

* Source/WebCore/rendering/updating/RenderTreeBuilderRuby.cpp:
(WebCore::RenderTreeBuilder::Ruby::findOrCreateParentForStyleBasedRubyChild):

Generate anonymous boxes for ruby so we get structure like

<ruby><ruby-base>base content</ruby-base><ruby-annotation>annotation content</ruby-annotation></ruby>

* Source/WebCore/rendering/updating/RenderTreeBuilderRuby.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::equivalentBlockDisplay):
(WebCore::Style::equivalentInlineDisplay):

Add inlinification function.

(WebCore::Style::shouldInlinifyForRuby):
(WebCore::Style::Adjuster::adjust const):

Inlinify children of ruby boxes.

* Source/WebCore/style/UserAgentStyle.cpp:
(WebCore::Style::UserAgentStyle::ensureDefaultStyleSheetsForElement):

Include ruby.css if enabled.

* Source/WebCore/style/UserAgentStyle.h:

Canonical link: https://commits.webkit.org/267937@main
